src/dev/mcp23017.h

```diff
@@ -0,0 +1,401 @@
+#pragma once
+
+#include "per/gpio.h"
+#include "per/i2c.h"
+
+namespace daisy
+{
+// Adapted from https://github.com/blemasle/arduino-mcp23017
+
+/**
+ * Registers addresses.
+ * The library use addresses for IOCON.BANK = 0.
+ * See "3.2.1 Byte mode and Sequential mode".
+ */
+enum class MCPRegister : uint8_t
+{
+    IODIR_A = 0x00, // Controls the direction of the data I/O for port A.
+    IODIR_B = 0x01, // Controls the direction of the data I/O for port B.
+    IPOL_A
+    = 0x02, // Configures the polarity on the corresponding GPIO_ port bits for port A.
+    IPOL_B
+    = 0x03, // Configures the polarity on the corresponding GPIO_ port bits for port B.
+    GPINTEN_A
+    = 0x04, // Controls the interrupt-on-change for each pin of port A.
+    GPINTEN_B
+    = 0x05, // Controls the interrupt-on-change for each pin of port B.
+    DEFVAL_A
+    = 0x06, // Controls the default comparaison value for interrupt-on-change for port A.
+    DEFVAL_B
+    = 0x07, // Controls the default comparaison value for interrupt-on-change for port B.
+    INTCON_A
+    = 0x08, // Controls how the associated pin value is compared for the interrupt-on-change for port A.
+    INTCON_B
+    = 0x09, // Controls how the associated pin value is compared for the interrupt-on-change for port B.
+    IOCON  = 0x0A, // Controls the device.
+    GPPU_A = 0x0C, // Controls the pull-up resistors for the port A pins.
+    GPPU_B = 0x0D, // Controls the pull-up resistors for the port B pins.
+    INTF_A = 0x0E, // Reflects the interrupt condition on the port A pins.
+    INTF_B = 0x0F, // Reflects the interrupt condition on the port B pins.
+    INTCAP_A
+    = 0x10, // Captures the port A value at the time the interrupt occured.
+    INTCAP_B
+    = 0x11, // Captures the port B value at the time the interrupt occured.
+    GPIO_A = 0x12, // Reflects the value on the port A.
+    GPIO_B = 0x13, // Reflects the value on the port B.
+    OLAT_A = 0x14, // Provides access to the port A output latches.
+    OLAT_B = 0x15, // Provides access to the port B output latches.
+};
+
+enum class MCPPort : uint8_t
+{
+    A = 0,
+    B = 1
+};
+
+inline MCPRegister operator+(MCPRegister a, MCPPort b)
+{
+    return static_cast<MCPRegister>(static_cast<uint8_t>(a)
+                                    + static_cast<uint8_t>(b));
+}
+
+enum class MCPMode : uint8_t
+{
+    INPUT,
+    INPUT_PULLUP,
+    OUTPUT,
+};
```

Comment:
I think these can all move inside of the primary device class instead of being global symbols (Unless there's a reason to leave them outside).

If you do move them inside the class we can remove the `MCP` prefix from each of the enum classes (i.e. `Register`, `Port`, `PinMode`) since they'll be scoped to the MCP23x17 class.

Reply:
Hmm, I think both `Mcp23017Transport` and `MCP23x17` classes use `MCPRegister` enum. I don't think I know the way to forward declare it. Any ideas?

Reply:
Oh, gotcha! No worries. I must have missed the shared usage when I was reading through it. 

Totally fine to leave it outside the class in that case.